### PR TITLE
[ALSP] Synchronization Engine `SyncRequest` spam detection (Permissionless-related engine level spam detection)

### PR DESCRIPTION
**Pull request by gomisha**
This PR adds ALSP spam level protection to the Synchronization Engine's `SyncRequest` message.

This is the first PR to implement ALSP misbehavior reporting at the engine level.

- Creates a probabilistic way to report ALSP misbehavior, based on a probability factor from 0 to 1 (e.g. 0.5 factor means it will report the behavior as spam about 50% of the time)
- Includes 2 load tests that test how many misbehavior reports are created based on different probability factors.
  - `TestLoad_Process_SyncRequest_HigherThanReceiver_OutsideTolerance_SometimesReportSpam`
  - `TestLoad_Process_SyncRequest_HigherThanReceiver_OutsideTolerance_AlwaysReportSpam`
- wires up checking for spam for each type of message type handled by `Process()` - these are all currently no-ops except for `SyncRequest`
- `SyncRequest` spam checking is hard coded to count a message as spam 1% of the time (i.e. approximately 1/100 messages); this will be updated in subsequent PRs to be configurable as a runtime flag when starting the engine

Future PRs will cover ALSP misbehavior reporting for other message types (e.g. RangeRequest, BatchRequest, SyncRequest, SyncResponse, BlockResponse)

ref: https://github.com/dapperlabs/flow-go/issues/6812

### CODE REVIEW

File: network/alsp/misbehavior.go [hunk 1]
```diff
@@ -43,6 +43,9 @@ const (
 
 	// UnauthorizedPublishOnChannel is a misbehavior that is reported when a message not authorized to be sent via pubsub is received via pubsub.
 	UnauthorizedPublishOnChannel network.Misbehavior = "unauthorized-pubsub-on-channel"
+
+	// PotentialSpam is a misbehavior that is reported when a message is potentially spam.
+	PotentialSpam network.Misbehavior = "potential-spam"
```
Comment (yhassanzadeh13) [inline network/alsp/misbehavior.go:48]: Please remove this new type and instead use the `ResourceIntensiveRequest` type for now. We will discuss the trajectory of the development in our next sync.

**Review by yhassanzadeh13 — APPROVED** [review]

File: engine/common/synchronization/engine.go [hunk 2]
```diff
@@ -202,9 +206,53 @@ func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, eve
 //   - All other errors are potential symptoms of internal state corruption or bugs (fatal).
 func (e *Engine) process(channel channels.Channel, originID flow.Identifier, event interface{}) error {
 	switch event.(type) {
-	case *messages.RangeRequest, *messages.BatchRequest, *messages.SyncRequest:
+	case *messages.BatchRequest:
+		report, misbehavior := e.validateBatchRequestForALSP(channel, originID, event)
+		if misbehavior {
+			e.con.ReportMisbehavior(report) // report misbehavior to ALSP
+			e.log.Warn().Hex("origin_id", logging.ID(originID)).Str(logging.KeySuspicious, "true").Msgf("received invalid batch request from %x: %v", originID[:], misbehavior)
```
[the remark this answers is not in the document]
Comment (gomisha) [reply engine/common/synchronization/engine.go]: 99857725d6d4f7002f141bbe79f97d59e84ead7e

File: engine/common/synchronization/engine.go [hunk 3]
```diff
@@ -202,9 +206,53 @@ func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, eve
...
+			e.metrics.InboundMessageDropped(metrics.EngineSynchronization, metrics.MessageBatchRequest)
+			return nil
+		}
+		return e.requestHandler.Process(channel, originID, event)
+	case *messages.RangeRequest:
+		report, misbehavior := e.validateRangeRequestForALSP(channel, originID, event)
+		if misbehavior {
+			e.con.ReportMisbehavior(report) // report misbehavior to ALSP
+			e.log.Warn().Hex("origin_id", logging.ID(originID)).Str(logging.KeySuspicious, "true").Msgf("received invalid range request from %x: %v", originID[:], misbehavior)
```
[the remark this answers is not in the document]
Comment (gomisha) [reply engine/common/synchronization/engine.go]: 99857725d6d4f7002f141bbe79f97d59e84ead7e

File: engine/common/synchronization/engine.go [hunk 4]
```diff
@@ -202,9 +206,53 @@ func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, eve
...
+			e.metrics.InboundMessageDropped(metrics.EngineSynchronization, metrics.MessageRangeRequest)
+			return nil
+		}
+		return e.requestHandler.Process(channel, originID, event)
+
+	case *messages.SyncRequest:
+		report, misbehavior := e.validateSyncRequestForALSP(channel, originID, event)
+		if misbehavior {
+			e.con.ReportMisbehavior(report) // report misbehavior to ALSP
+			e.log.Warn().Hex("origin_id", logging.ID(originID)).Str(logging.KeySuspicious, "true").Msgf("received invalid sync request from %x: %v", originID[:], misbehavior)
```
[the remark this answers is not in the document]
Comment (gomisha) [reply engine/common/synchronization/engine.go]: 99857725d6d4f7002f141bbe79f97d59e84ead7e

File: engine/common/synchronization/engine.go [hunk 5]
```diff
@@ -202,9 +206,53 @@ func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, eve
...
+			return nil
+		}
 		return e.requestHandler.Process(channel, originID, event)
-	case *messages.SyncResponse, *messages.BlockResponse:
+
+	case *messages.BlockResponse:
+		report, misbehavior := e.validateBlockResponseForALSP(channel, originID, event)
+		if misbehavior {
+			e.con.ReportMisbehavior(report) // report misbehavior to ALSP
+			e.log.Warn().Hex("origin_id", logging.ID(originID)).Str(logging.KeySuspicious, "true").Msgf("received invalid block response from %x: %v", originID[:], misbehavior)
```
[the remark this answers is not in the document]
Comment (gomisha) [reply engine/common/synchronization/engine.go]: 99857725d6d4f7002f141bbe79f97d59e84ead7e

File: engine/common/synchronization/engine.go [hunk 6]
```diff
@@ -202,9 +206,53 @@ func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, eve
...
+			e.metrics.InboundMessageDropped(metrics.EngineSynchronization, metrics.MessageBlockResponse)
+			return nil
+		}
+		return e.responseMessageHandler.Process(originID, event)
+
+	case *messages.SyncResponse:
+		report, misbehavior := e.validateSyncResponseForALSP(channel, originID, event)
+		if misbehavior {
+			e.con.ReportMisbehavior(report) // report misbehavior to ALSP
+			e.log.Warn().Hex("origin_id", logging.ID(originID)).Str(logging.KeySuspicious, "true").Msgf("received invalid sync response from %x: %v", originID[:], misbehavior)
```
[the remark this answers is not in the document]
Comment (gomisha) [reply engine/common/synchronization/engine.go]: 99857725d6d4f7002f141bbe79f97d59e84ead7e

File: engine/common/synchronization/engine.go [hunk 7]
```diff
@@ -424,3 +472,59 @@ func (e *Engine) sendRequests(participants flow.IdentifierList, ranges []chainsy
...
+// validateSyncRequestForALSP checks if a sync request should be reported as spam. It returns a misbehavior report and a boolean indicating whether the request should be reported.
+func (e *Engine) validateSyncRequestForALSP(channel channels.Channel, originID flow.Identifier, event interface{}) (*alsp.MisbehaviorReport, bool) {
+	// Generate a random integer between 1 and spamProbabilityMultiplier (exclusive)
+	n, err := rand.Uint32n(spamProbabilityMultiplier)
+
+	if err != nil {
+		// failing to generate a random number is unlikely. If an error is encountered while
+		// generating a random number it indicates a bug and processing can not proceed.
+		e.log.Fatal().
+			Err(err).
```
[the remark this answers is not in the document]
Comment (gomisha) [reply engine/common/synchronization/engine.go:520]: 1aa198a940b4f8a63bff3b722253932820988817

File: engine/common/synchronization/engine.go [hunk 8]
```diff
@@ -424,3 +472,59 @@ func (e *Engine) sendRequests(participants flow.IdentifierList, ranges []chainsy
...
+	if float32(n) < e.spamReportConfig.syncRequestProbability*spamProbabilityMultiplier {
+		// create a misbehavior report
+		e.log.Info().Str("originID", originID.String()).Msg("creating misbehavior report")
+		report, err := alsp.NewMisbehaviorReport(originID, alsp.PotentialSpam)
+
+		if err != nil {
+			// failing to create the misbehavior report is unlikely. If an error is encountered while
+			// creating the misbehavior report it indicates a bug and processing can not proceed.
+			e.log.Fatal().
+				Err(err).
```
[the remark this answers is not in the document]
Comment (gomisha) [reply engine/common/synchronization/engine.go:537]: 1aa198a940b4f8a63bff3b722253932820988817

File: network/alsp/misbehavior.go (hunk 1 above)
Comment (gomisha) [reply network/alsp/misbehavior.go]: 2967f4a70a86560c9265ebf56559dbf6bf66166d

File: engine/common/synchronization/engine.go [hunk 9]
```diff
@@ -105,6 +108,7 @@ func New(
 		pollInterval:         opt.PollInterval,
 		scanInterval:         opt.ScanInterval,
 		participantsProvider: participantsProvider,
+		spamReportConfig:     NewSpamReportConfig(),
```
[the remark this answers is not in the document]
Comment (gomisha) [reply engine/common/synchronization/engine.go]: bdb589999a738c627f17278519647aafb3174b0b

**Review by kc1116 — APPROVED** [review]

**Pull request merged**